pkg/controller/operators/catalog/og_source_provider.go

```diff
@@ -0,0 +1,106 @@
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+type OperatorGroupToggleSourceProvider struct {
+	sp              SourceProviderWithInvalidate
+	logger          *logrus.Logger
+	ogLister        v1listers.OperatorGroupLister
+	globalNamespace string
+}
+
+func NewOperatorGroupToggleSourceProvider(sp SourceProviderWithInvalidate, logger *logrus.Logger,
+	ogLister v1listers.OperatorGroupLister, globalNamespace string) *OperatorGroupToggleSourceProvider {
+	return &OperatorGroupToggleSourceProvider{
+		sp:              sp,
+		logger:          logger,
+		ogLister:        ogLister,
+		globalNamespace: globalNamespace,
+	}
+}
+
+// SourceProviderWithInvalidate is a client-side interface for the OperatorGroup Toggle
+type SourceProviderWithInvalidate interface {
+	cache.SourceProvider
+	Invalidate(key cache.SourceKey)
+}
+
+const exclusionAnnotation string = "olm.operatorframework.io/exclude-global-namespace-resolution"
+
+func (e *OperatorGroupToggleSourceProvider) Sources(namespaces ...string) map[cache.SourceKey]cache.Source {
+	// Check if annotation is set first
+	resolutionNamespaces, err := e.CheckForExclusion(namespaces...)
+	if err != nil {
+		// Fail early with a dummy Source that returns an error
+		// Note: this only errors in the case the Lister fails, which should not happen
+		// TODO: Update the Sources interface to return an error
+		m := make(map[cache.SourceKey]cache.Source)
+		key := cache.SourceKey{Name: "og-exclusion-failure", Namespace: namespaces[0]}
+		source := errorSource{err}
+		m[key] = source
+		return m
+	}
+	return e.sp.Sources(resolutionNamespaces...)
+}
+
+func (e *OperatorGroupToggleSourceProvider) Invalidate(key cache.SourceKey) {
+	e.sp.Invalidate(key)
+}
+
+type errorSource struct {
+	error
+}
+
+func (e errorSource) Snapshot(ctx context.Context) (*cache.Snapshot, error) {
+	return nil, e.error
+}
+
+func (e *OperatorGroupToggleSourceProvider) CheckForExclusion(namespaces ...string) ([]string, error) {
+	var defaultResult = namespaces
+	// The first namespace provided is always the current namespace being synced
+	var ownNamespace = namespaces[0]
+	var toggledResult = []string{ownNamespace}
+
+	if ownNamespace == e.globalNamespace {
+		// Global namespace is being synced
+		// Return early with default
+		return defaultResult, nil
+	}
+
+	// Check the OG on the NS provided for the exclusion annotation
+	ogs, err := e.ogLister.OperatorGroups(ownNamespace).List(labels.Everything())
+	if err != nil {
+		// Assume resolution was global in the case of an error (the default behavior)
+		return defaultResult, fmt.Errorf("listing operatorgroups in namespace %s: %s", ownNamespace, err)
+	}
+
+	if len(ogs) != 1 || ogs[0] == nil {
+		// Problem with the operatorgroup configuration in the namespace
+		// The namespace may not be configured as an operator installation target namespace
+		// Assume global resolution (default)
+		return defaultResult, nil
```

Comment:
This could give you flaky behavior in cases where users are simultaneously creating an OperatorGroup with Subscriptions -- if the catalog-operator sees one of the Subscriptions before it sees the OperatorGroup, it could end up doing the wrong thing. Better off returning an error here and letting it try again during the namespace resync triggered by seeing the OperatorGroup.

Reply:
I agree there's a race condition, but would returning an error here change the default behavior of OLM? Where now creating a subscription in a namespace without 1 OG in it is a resolution-time error, instead of a runtime error. Is it safe to make that change?

Reply:
I'm not sure exactly it fails today. I do know that there has occasionally been talk about respecting supported install modes in catalog-operator decisions, and I think you can make a good argument for preventing the creation of an InstallPlan if it's known to be invalid ahead of time. In this particular situation, the (hypothetical) InstallPlan may or may not be invalid, and the user may or may not have wanted to exclude operators from the global catalogs.

Is there a realistic use case for creating an InstallPlan in a namespace that doesn't have an OperatorGroup? It would have to be pretty convincing to accept that decisions may vary when creating OGs and Subscriptions at about the same time.